src/context.c

```diff
@@ -290,6 +291,171 @@ ly_ctx_new(const char *search_dir, uint16_t options, struct ly_ctx **new_ctx)
+            ly_set_free(set, free);
+            return LY_EINVAL;
+        }
+    }
+
+    ly_set_free(set, free);
+    return LY_SUCCESS;
+}
+
+static LY_ERR
+ly_ctx_new_yl_common(const char *search_dir, const char *input, LYD_FORMAT format, int options,
+        LY_ERR (*parser_func)(const struct ly_ctx *, const char *, LYD_FORMAT, uint32_t, uint32_t, struct lyd_node **),
+        struct ly_ctx **ctx)
+{
+    struct lyd_node *module, *node;
+    struct lyd_node *yltree = NULL;
+    struct ly_set *set = NULL;
+    const char **feature_arr = NULL;
+    struct ly_set features = {0};
+    LY_ERR ret = LY_SUCCESS;
+    int err = 0;
+
+    /* create empty (with internal modules including ietf-yang-library) context */
+    LY_CHECK_GOTO(ret = ly_ctx_new(search_dir, options, ctx), error);
+
+    /* parse yang library data tree */
+    LY_CHECK_GOTO(ret = parser_func(*ctx, input, format, 0, LYD_VALIDATE_PRESENT, &yltree), error);
+
+    LY_CHECK_GOTO(ret = lyd_find_xpath(yltree, "/ietf-yang-library:yang-library/module-set[1]/module", &set), error);
+
+    if (set->count == 0) {
+        /* perhaps a legacy data tree? */
+        LY_CHECK_GOTO(ret = ly_ctx_new_yl_legacy(ctx, yltree), error);
+    } else {
+        /* process the data tree */
+        for (uint32_t i = 0; i < set->count; ++i) {
+            module = set->dnodes[i];
+
+            /* initiate */
+            const char *name = NULL, *revision = NULL;
+
+
+            /* Iterate over data */
+            LY_LIST_FOR(*lyd_node_child_p(module), node) {
+                if (!strcmp(node->schema->name, "name")) {
+                    name = ((struct lyd_node_term *)node)->value.canonical;
+                } else if (!strcmp(node->schema->name, "revision")) {
+                    revision = ((struct lyd_node_term *)node)->value.canonical;
+                } else if (!strcmp(node->schema->name, "feature")) {
+                    LY_CHECK_GOTO(ret = ly_set_add(&features, node, 0, NULL), error);
+                }
+            }
+
+            feature_arr = malloc(sizeof(char *) * (features.count + 1));
+            LY_CHECK_ERR_GOTO(feature_arr == NULL, ret = LY_EMEM, error);
+
+            /* Parse features into an array of strings */
+            for (uint32_t u = 0; u < features.count; u++) {
+                feature_arr[u] = ((struct lyd_node_term *)features.dnodes[u])->value.canonical;
+            }
+            feature_arr[features.count] = NULL;
+            ly_set_clean(&features, NULL);
+
+            /* use the gathered data to load the module */
+            const struct lys_module *mod = ly_ctx_load_module(*ctx, name, revision, feature_arr);
+            free(feature_arr);
+            feature_arr = NULL;
+            if (!mod) {
+                LOGERR(NULL, LY_EINVAL, "Unable to load module specified by yang library data.");
+                ret = LY_EINVAL;
+                goto error;
+            }
+        }
+    }
+
+    if (0) {
```

Comment:
Nope, never write something like this again. I believe the whole `err` variable is useless because you can simply use `ret` - `!ret` - success, `ret` - error.